src/meta/src/hummock/compaction_schedule_policy.rs

```diff
@@ -0,0 +1,722 @@
+            }
+        }
+
+        if let Some((pending_bytes, context_id)) = next_idle_compactor {
+            let compactor = self.update_compactor(context_id, pending_bytes, compact_task, true);
+            Some(compactor)
+        } else {
+            None
+        }
+    }
+
+    fn next_compactor(&mut self, compact_task: Option<&CompactTask>) -> Option<Arc<Compactor>> {
+        let mut next_compactor = None;
+        if let Some((pending_bytes, context_id)) = self
+            .pending_bytes_to_compactor
+            .keys()
+            .min_by(|(pb1, _), (pb2, _)| pb1.cmp(pb2))
+        {
+            next_compactor = Some((*pending_bytes, *context_id));
+        }
+
+        if let Some((pending_bytes, context_id)) = next_compactor {
+            let compactor = self.update_compactor(context_id, pending_bytes, compact_task, true);
+            Some(compactor)
+        } else {
+            None
+        }
+    }
+
+    fn random_compactor(&mut self, compact_task: Option<&CompactTask>) -> Option<Arc<Compactor>> {
+        if self.pending_bytes_to_compactor.is_empty() {
+            return None;
+        }
+
+        let compactor_index =
+            rand::thread_rng().gen::<usize>() % self.compactor_to_pending_bytes.len();
+        // `BTreeMap` does not record subtree size, so O(logn) method to find the nth smallest
+        // element is not available.
+        let (context_id, pending_bytes) = self
+            .compactor_to_pending_bytes
+            .iter()
+            .nth(compactor_index)
+            .unwrap();
+
+        let compactor = self.update_compactor(*context_id, *pending_bytes, compact_task, true);
+        Some(compactor)
+    }
+
+    fn add_compactor(
+        &mut self,
+        context_id: HummockContextId,
+        max_concurrent_task_number: u64,
+    ) -> Receiver<MetaResult<SubscribeCompactTasksResponse>> {
+        let (tx, rx) = tokio::sync::mpsc::channel(STREAM_BUFFER_SIZE);
+        if let Some(pending_bytes) = self.compactor_to_pending_bytes.remove(&context_id) {
+            self.pending_bytes_to_compactor
+                .remove(&(pending_bytes, context_id))
+                .unwrap();
+        }
+        self.pending_bytes_to_compactor.insert(
+            (0, context_id),
+            Arc::new(Compactor::new(context_id, tx, max_concurrent_task_number)),
+        );
+        self.compactor_to_pending_bytes.insert(context_id, 0);
+        tracing::info!("Added compactor session {}", context_id);
+        rx
+    }
+
+    fn remove_compactor(&mut self, context_id: HummockContextId) {
+        if let Some(pending_bytes) = self.compactor_to_pending_bytes.remove(&context_id) {
+            self.pending_bytes_to_compactor
+                .remove(&(pending_bytes, context_id))
+                .unwrap();
+        }
+        tracing::info!("Removed compactor session {}", context_id);
+    }
+
+    fn report_compact_task(&mut self, context_id: HummockContextId, task: &CompactTask) {
```

Comment:
Could we check `TaskStatus` instead of using `assign_task` boolean ?

Reply:
Good idea.